### DISCUSSION TITLE
Passing context as second argument to onReceive callback

### DISCUSSION THREAD

**Discussion by matigda** [statelyai/xstate#1554] [Q&A]
Hi there!
First of all I want to say that this is awesome tool! I'm working with it for a few months now and it really helped me solve some issues.

There is one design thing though that I don't really understand and it's making my life harder.

I have service that looks like that:

 ```
itemsService: (context: CheckoutContextSchema<ShippingMethodsType>, event: CheckoutEventSchema) => (cb: (event: object) => void, onReceive: any) => {
            onReceive(async (event: CheckoutEventSchema) => {

                if (event.type === EventId.ADD_PRODUCT_ON_BACKEND) {
                    let cartId: string;
                    if (!context.cartId) {
                        cartId = await createCart();
                        cb({type: EventId.ASSIGN_CART_ID, payload: {cartId}})
                    } else {
                        cartId = context.cartId
                    }

                    await addProductToCart(event.payload.product, event.payload.quantity, cartId);
                    cb({type: EventId.PRODUCT_ADDED});
                }
            })
        },
```

The issue here is that context is being used in a closure, so it never changes. So even though this line `cb({type: EventId.ASSIGN_CART_ID, payload: {cartId}})` sets the cart id in context, this condition: `if (!context.cartId) {` always returns true. Can you tell me why context is not passed to `onReceive` function?

**Comment by davidkpiano** [comment]
> The issue here is that context is being used in a closure, so it never changes.

Yes, that's the point. When you create the invoked callback (with the `context` at the time of its creation), the only way you can communicate between the parent/child is with _events_.

So, in order for this actor to receive some updated data, it needs to receive that data in an event (via `onReceive`).
